contracts/infrastructure/DappRegistry.sol

```diff
@@ -2,77 +2,197 @@ pragma solidity ^0.6.12;
 import "./storage/Storage.sol";
 
-contract DappRegistry is IAuthoriser, Storage, Owned {
+contract DappRegistry is IAuthoriser, Storage {
 
-    bytes32 constant internal ARGENT_DEFAULT = keccak256("argent_default");
+    // The timelock period
+    uint64 public securityPeriod;
+    // The new timelock period
+    uint64 public newSecurityPeriod;
+    // Time at which the new timelock becomes effective
+    uint64 public securityPeriodChangeAfter;
 
-    struct Authorisation {
-        bool isActive;
-        address filter;
-    }
+    // bit vector of enabled registry ids for each wallet
+    mapping (address => bytes32) public enabledRegistryIds; // [wallet] => [bit vector of 256 registry ids]
+    // authorised dapps and their filters for each registry id
+    mapping (uint8 => mapping (address => bytes32)) public authorisations; // [registryId] => [dapp] => [{filter:160}{validAfter:64}]
+    // pending authorised dapps and their filters for each registry id
+    mapping (uint8 => mapping (address => bytes32)) public pendingFilterUpdates; // [registryId] => [dapp] => [{filter:160}{validAfter:64}]
+    // owners for each registry id
+    mapping (uint8 => address) public registryOwners; // [registryId] => [owner]
     
-    // mapping [wallet][name][status]
-    mapping (address => mapping (bytes32 => bool)) public isEnabled;
-    // mapping [name][address][authorisation]
-    mapping (bytes32 => mapping (address => Authorisation)) public authorisations;
+    event RegistryCreated(uint8 registryId, address registryOwner);
+    event RegistryRemoved(uint8 registryId);
+    event OwnerChanged(uint8 registryId, address newRegistryOwner);
+    event TimelockChangeRequested(uint64 newSecurityPeriod);
+    event TimelockChanged(uint64 newSecurityPeriod);
+    event FilterUpdated(uint8 indexed registryId, address dapp, address filter, uint256 validFrom);
+    event FilterUpdateRequested(uint8 indexed registryId, address dapp, address filter, uint256 validFrom);
+    event DappAdded(uint8 indexed registryId, address dapp, address filter, uint256 validFrom);
+    event DappRemoved(uint8 indexed registryId, address dapp);
 
-    bytes32[] public registries;
+    modifier onlyOwner(uint8 _registryId) {
+        validateOwner(_registryId);
+        _;
+    }
+    
+    constructor(uint64 _securityPeriod) public {
+        // set the timelock period
+        securityPeriod = _securityPeriod;
+        // set the owner of the Argent Registry (registryId == 0)
+        registryOwners[0] = msg.sender;
+    }
 
-    mapping (bytes32 => address) public registryManager;
+    /********* Wallet-centered functions *************/
 
-    event RegistryCreated(bytes32 registry, address manager);
-    event RegistryRemoved(bytes32 registry);
+    function isEnabledRegistry(address _wallet, uint8 _registryId) external view returns (bool) {
+        uint registries = uint(enabledRegistryIds[_wallet]);
+        return ((registries >> _registryId) & 1) > 0;
```

Comment:
Remix does something odd and allows this to overflow, i.e. this returns `true` there for a registryId > 255. I'm pretty sure this can't be the case with the right shift here but putting a note to add a test for that.

Reply:
Looks to me like solidity uses `_registryId % 256` for `_registryId` when the function is called with `_registryId` > 256. So it's expected that `isEnabledRegistry(address _wallet, uint8 _registryId)` is going to return true for `uint` values of `_registryId` such that `isEnabledRegistry(_wallet, _registryId % 256) == true`. I verified that in a test:
```
const dataWithBadSig = utils.encodeFunctionCall("isEnabledRegistry(address,uint)", [wallet.address, 256 + CUSTOM_REGISTRY_ID]);
      const correctSig = web3.utils.sha3("isEnabledRegistry(address,uint8)").slice(0, 10);
      const data = `${correctSig}${dataWithBadSig.slice(10)}`;
      console.log({ dataWithBadSig, correctSig, data });
      const output = await web3.eth.call({
        to: authoriser.address,
        data
      });
      assert.equal(output.toString(), "0x0000000000000000000000000000000000000000000000000000000000000001");
```

Reply:
You mean that it's allowed to overflow?

Reply:
Yes in the sense that the transaction won't revert if your "uint8" is > 255. It will just "wrap around" (i.e. use the value passed modulo 256).